sdk/dotnet/Pulumi/Deployment/TaskMonitoringHelper.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Pulumi
+{
+    /// Monitors dynamically added tasks for completion. Enters IDLE
+    /// state when all monitored tasks finish. Allows awaiting next
+    /// IDLE state or the first exception, whichever comes first.
+    /// Thread-safe.
+    internal class TaskMonitoringHelper
+    {
+        private readonly object _lockObject = new Object();
+        private int _activeTasks = 0;
+        private TaskCompletionSource<Exception?>? _next = null;
+
+        public TaskMonitoringHelper()
+        {
+            CheckInvariants();
+        }
+
+        /// Adds the task to the task-set being monitored.
+        public void AddTask(Task task)
+        {
+            lock (_lockObject)
+            {
+                // Check if we are moving out of IDLE state.
+                if (_activeTasks == 0)
+                {
+                    // Fresh promise is needed.
+                    _next = new TaskCompletionSource<Exception?>();
+                }
+                _activeTasks++;
+                CheckInvariants();
+            };
+            task.ContinueWith(this.OnTaskCompleted);
```

Comment:
This comment makes me smile because I really am lucky to have a teammate that cares about allocations, while also being a skeptic this will ever matter (if we put a counter on Tasks allocated in pulumi programs I doubt it would ever reach 1mln?), it's just fun.